torch_xla/csrc/ops/ops.cpp

```diff
@@ -566,6 +567,32 @@ torch::lazy::NodePtr Norm(const torch::lazy::Value& input,
   return Pow(result, norm_exp_inv);
 }
 
+torch::lazy::NodePtr Pdist_forward(const torch::lazy::Value& input,
+                                   const c10::optional<at::Scalar>& p,
+                                   c10::optional<at::ScalarType> dtype) {
+  // pdist(x, p) is equal to norm(x[:, None]-x, dim=2, p) and we only take the
+  // upper triangle without diagonal line.
+  auto lower_fn = [=](const XlaNode& node,
+                      LoweringContext* loctx) -> XlaOpVector {
+    xla::XlaOp xla_input = loctx->GetOutputOp(node.operand(0));
+    return node.ReturnOp(BuildUpperTriangle(xla_input), loctx);
+  };
+  auto lower_for_shape_fn =
+      [&](absl::Span<const xla::XlaOp> operands) -> xla::XlaOp {
+    return BuildUpperTriangle(operands[0]);
+  };
+  torch::lazy::NodePtr tmp = input - torch::lazy::MakeNode<Unsqueeze>(input, 1);
+  torch::lazy::NodePtr result_matrix = Norm(tmp, p, dtype, {2}, false);
```

Comment:
The logic LGTM, but typically we organize this logic in such a way that we have a function called `BuildPdistForward` (instead of `BuildUpperTriangle`) to have all the lowering-specific logic in that function instead of here. Can we do that for this as well?

Reply:
Understood, makes sense. I was referring to putting the lines 584-585 in that `BuildPdistForward` as well.

Reply:
I see, this case is a bit tricky. Line 584-585 were handled in node(`torch::lazy::NodePtr`) level using the existing `Unsqueeze` and `Norm` Nodes, while the remaining is handled in lower xla op level since they don't provide `Norm` etc. as we needed here. I didn't see others handle `torch::lazy::NodePtr` in `Build*` functions. That's why I split them out.

Reply:
I see, sg, let's just leave it as is. Thanks!